changelog/comma-deprecation-error.dd

```diff
@@ -0,0 +1,59 @@
+The deprecation period of using the result of comma expression has ended
+
+Comma expressions have proven to be a frequent source of confusion, and bugs.
+Using their result will now trigger an error message.
+
+$(P The comma operator (`,`) allows executing multiple expressions and
+    discards the result of them except for the last which is returned.
+
+    ---
+    int a = 1;
+    int b = 2;
+    bool ret = a == 2, b == 2; // true
+    ---
+
+    It's also common to use the comma operator in for-loop increment
+    statements to allow multiple expressions.
+
+    ---
+    for (; !a.empty && !b.empty; a.popFront, b.popFront)
+    ---
+)
+$(H4 Corrective Action)
+$(P If possible, split the comma operator in two statements. Otherwise use
+    lambdas.
+
+    ---
+    auto result = foo(), bar();
+
+    // split off in two statements
+    foo();
+    auto result = bar();
+
+    // or use lambdas
+    auto result = {foo(); return bar();}();
+    ---
+)
+$(H4 Rationale)
+$(P The comma operator leads to unintended behavior (see below for a selection)
+    Moreover it is not commonly used and it blocks the ability to implement tuples.
+
+    A selection of problems through the accidental use of the comma operator:
+    ---
+    writeln( 6, mixin("7,8"), 9 ); // 6, 8, 9
+
+    template vec(T...)(T args) { ... }
+    vec v = (0, 0, 3); // 3, because vec is variadic
+
+    int b = 2;
+    if (a == 1, b == 2) {
+        // will always be reached
+    }
+
+    void foo(int x) {}
+    foo((++a, b));
+
+    synchronized (lockA, lockB) {}
+    // isn't currently implemented, but still compiles "thanks" to the comma operator
```

Comment:
`"thanks"` sounds a bit condescending. Perhaps replace with `due`.

Reply:
Also submitted for dlang.org: https://github.com/dlang/dlang.org/pull/2152